lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
 }
 
+static std::optional<uint64_t> electFSO(const auto &MaybeReturns) {
+  auto It = std::min_element(MaybeReturns.begin(),
+                             MaybeReturns.end(),
+                             [](const auto &LHS, const auto &RHS) {
+                               return LHS.second < RHS.second;
+                             });
+  if (It == MaybeReturns.end())
+    return {};
+  return It->second;
+}
+
+static UpcastablePointer<model::FunctionEdge>
+makeEdge(MetaAddress Destination, model::FunctionEdgeType::Values Type) {
+  model::FunctionEdge *Result = nullptr;
+  if (model::FunctionEdgeType::isCall(Type))
+    Result = new model::CallEdge(Destination, Type);
+  else
+    Result = new model::FunctionEdge(Destination, Type);
+  return UpcastablePointer<model::FunctionEdge>(Result);
+};
+
+template<class FunctionOracle>
+SortedVector<model::BasicBlock>
+CFEPAnalyzer<FunctionOracle>::collectDirectCFG(llvm::BasicBlock *Entry,
+                                               llvm::Function *F) {
+  using namespace llvm;
+  SortedVector<model::BasicBlock> CFG;
+
+  for (auto &BB : *F) {
+    if (GCBI::isJumpTarget(&BB)) {
+      auto Start = getBasicBlockPC(&BB);
+      MetaAddress End;
+      uint64_t Size;
+
+      bool Stop = false;
+      model::BasicBlock Block{ Start };
+      OnceQueue<BasicBlock *> Queue;
+      Queue.insert(&BB);
+
+      while (!Queue.empty()) {
+        BasicBlock *Current = Queue.pop();
+
+        auto [CandidateEnd, CandidateSize] = getPC(Current->getTerminator());
+        if (CandidateEnd > End) {
+          End = CandidateEnd;
+          Size = CandidateSize;
+        }
+
+        for (auto *Succ : successors(Current)) {
+          if (GCBI::isJumpTarget(Succ)) {
+            auto Destination = getBasicBlockPC(Succ);
+            auto Edge = makeEdge(Destination,
+                                 model::FunctionEdgeType::DirectBranch);
+            Block.Successors.insert(Edge);
+          } else if (UnexpectedPCCloned == Succ && succ_size(Current) == 1) {
+            Block.Successors.insert(makeEdge(MetaAddress::invalid(),
+                                             model::FunctionEdgeType::LongJmp));
+          } else {
+            auto *I = Succ->getFirstNonPHI();
+
+            if (isa<CallInst>(I)
+                && cast<CallInst>(I)->getCalledFunction() == PreHookMarker) {
+              MetaAddress Destination;
+              model::FunctionEdgeType::Values Type;
+              auto *CalleePC = cast<CallInst>(I)->getArgOperand(1);
+
+              // Direct or indirect call?
+              if (isa<ConstantStruct>(CalleePC)) {
+                Destination = MetaAddress::fromConstant(CalleePC);
+                Type = model::FunctionEdgeType::FunctionCall;
+              } else {
+                Destination = MetaAddress::invalid();
+                Type = model::FunctionEdgeType::IndirectCall;
+              }
+
+              auto Edge = makeEdge(Destination, Type);
```

Comment:
You're using `auto` too much. 
Use `auto` only if:
1. The line is very long and you can't figure any other way to make `check-conventions` happy.
2. The type is obvious since it's expressedly stated on the same line (e.g., `auto *Lol = dyn_cast<CallInst>(...)`).
3. The type is very ugly (typically, iterators).
Please review all of the usages in this sense.

Reply:
OK, I have reviewed basically all the instances here. I've basically kept `auto` only as result type of `makeEdge`, which is an `UpcastablePointer<T>` and I need `auto` here to pass `check-revng-conventions`.